FirebaseMLModelDownloader/Sources/ModelInfoRetriever.swift

```diff
@@ -39,74 +56,47 @@ enum DownloadModelInfoResult {
+  private let apiKey: String
   /// Current Firebase app name.
   private let appName: String
   /// Local model info to validate model freshness.
-  private var localModelInfo: LocalModelInfo?
+  private let localModelInfo: LocalModelInfo?
 
   /// Associate model info retriever with current Firebase app, and model name.
-  init(modelName: String, options: FirebaseOptions, installations: Installations, appName: String,
-       localModelInfo: LocalModelInfo? = nil) {
+  init(modelName: String,
+       projectID: String,
+       apiKey: String,
+       installations: Installations,
+       appName: String,
+       localModelInfo: LocalModelInfo? = nil,
+       session: ModelInfoRetrieverSession? = nil) {
     self.modelName = modelName
-    self.options = options
+    self.projectID = projectID
+    self.apiKey = apiKey
     self.installations = installations
     self.appName = appName
     self.localModelInfo = localModelInfo
-  }
-}
-
-/// Extension to handle fetching model info from server.
-extension ModelInfoRetriever {
-  /// HTTP request headers.
-  private static let fisTokenHTTPHeader = "x-goog-firebase-installations-auth"
-  private static let hashMatchHTTPHeader = "if-none-match"
-  private static let bundleIDHTTPHeader = "x-ios-bundle-identifier"
-
-  /// HTTP response headers.
-  private static let etagHTTPHeader = "Etag"
-
-  /// Construct model fetch base URL.
-  var modelInfoFetchURL: URL {
-    let projectID = options.projectID ?? ""
-    let apiKey = options.apiKey
-    var components = URLComponents()
-    components.scheme = "https"
-    components.host = "firebaseml.googleapis.com"
-    components.path = "/v1beta2/projects/\(projectID)/models/\(modelName):download"
-    components.queryItems = [URLQueryItem(name: "key", value: apiKey)]
-    // TODO: handle nil
-    return components.url!
-  }
-
-  /// Construct model fetch URL request.
-  private func getModelInfoFetchURLRequest(token: String) -> URLRequest {
-    var request = URLRequest(url: modelInfoFetchURL)
-    request.httpMethod = "GET"
-    // TODO: Check if bundle ID needs to be part of the request header.
-    let bundleID = Bundle.main.bundleIdentifier ?? ""
-    request.setValue(bundleID, forHTTPHeaderField: ModelInfoRetriever.bundleIDHTTPHeader)
-    request.setValue(token, forHTTPHeaderField: ModelInfoRetriever.fisTokenHTTPHeader)
-    /// Get model hash if local model info is available on device.
-    if let modelInfo = localModelInfo {
-      request.setValue(
-        modelInfo.modelHash,
-        forHTTPHeaderField: ModelInfoRetriever.hashMatchHTTPHeader
-      )
+    if let urlSession = session {
+      self.session = urlSession
+    } else {
+      self.session = URLSession(configuration: .ephemeral)
     }
-    return request
   }
 
   /// Get installations auth token.
-  private func getAuthToken(completion: @escaping (Result<String, DownloadError>) -> Void) {
+  lazy var authToken = { (completion: @escaping (Result<String, DownloadError>) -> Void) in
```

Comment:
As discussed offline, this property probably should be renamed to something like `authTokenProvider` or `authTokenRetriever`. Also we most likely need to remove `installations` property from the class and use the Installations static method in the real `authTokenProvider (we may need to pass FIRApp as a param along with `completion` into the closure to get installation instance for the specific app).

Reply:
Good catch, I renamed it to `authTokenProvider`. However I would prefer not to pass `FirebaseApp` past `ModelDownloader` if that's possible. Ideally, it would be best if I could pass the entire definition of `authTokenProvider` from `ModelDownloader` but I haven't had a chance to figure out the signature of that parameter. What are your thoughts?

Reply:
Resolved with the changes in #7355.